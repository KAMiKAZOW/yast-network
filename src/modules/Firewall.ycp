/**
 * File:	modules/Firewall.ycp
 * Package:	Network configuration
 * Summary:	Firewall data
 * Authors:	Michal Svec <msvec@suse.cz>
 *
 * $Id$
 */

{

module "Firewall";
textdomain "network";

import "Runlevel";

/**
 * Firewalled devices
 */
list devices = [];
list devices_init = [];

/* NI: required wrt functions in routines.ycp */
block<boolean> AbortFunction = nil;
define boolean Modified() ``{ return true; }

include "network/routines.ycp";
include "network/runtime.ycp";

/**
 * Read firewall settings
 * @return true if success
 */
global define boolean Read() ``{

    /* read firewalled devices */
    string devs = "";
    if(SCR::Read(.target.size, "/etc/sysconfig/personal-firewall") > 0)
	devs = (string) SCR::Read(.sysconfig.personal-firewall.REJECT_ALL_INCOMING_CONNECTIONS);
    y2debug("devs=%1", devs);
    if(devs == nil) return false;
    if(devs == "") return true;

    devices = splitstring(devs, " ");
    y2debug("devices=%1", devices);
    devices_init = (list) eval(devices);

    return true;
}

/**
 * Write firewall settings and apply changes
 * @return true if success
 */
global define boolean Write() ``{

    /* Check if there is anything to do */
    if(devices_init == devices) {
	y2debug("Devices not modified");
	return true;
    }

    /* Update sysconfig */
    string devs = "";
    y2debug("devices=%1", devices);
    devices = filter(`v, devices, ``(v != "no"));
    devices = filter(`v, devices, ``(v != "nothing"));
    y2debug("devices=%1", devices);
    if(size(devices) > 0) devs = mergestring(devices, " ");
    SCR::Write(.sysconfig.personal-firewall.REJECT_ALL_INCOMING_CONNECTIONS, devs);
    SCR::Write(.sysconfig.personal-firewall, nil);

    /* Setup init scripts */
    if(size(devices) > 0) {
	EnableNetwork();
	EnableService("SuSEfirewall2_init");
	EnableService("SuSEfirewall2_setup");
	EnableService("SuSEfirewall2_final");
	SCR::Execute(.target.bash, "/sbin/SuSEfirewall2 start");
    }
    else {
	DisableService("SuSEfirewall2_final");
	DisableService("SuSEfirewall2_setup");
	DisableService("SuSEfirewall2_init");
	SCR::Execute(.target.bash, "/sbin/SuSEfirewall2 stop");
    }

    return true;
}

/**
 * Get all the firewall configuration from a map.
 * When called by *_auto (preparing autoinstallation data)
 * the list may be empty.
 * @param settings autoinstallation settings
 * @return true if success
 */
global define boolean Import(list settings) ``{
    if(size(settings) == 0)
	return true;

    devices = (list) eval(settings);
    return true;
}

/**
 * Dump the firewall settings to a list, for autoinstallation use.
 * @return autoinstallation settings
 */
global define list Export() ``{
    return (list) eval(devices);
}

/**
 * Convert real device to the firewall device specification
 * @param device real device
 * @return firewall device
 */
define string device2fw(string device) ``{
    if(device == nil || device == "") {
	y2error("wrong device identification: %1", device);
	return nil;
    }
    if(device_type(device) == "ppp") return "modem";
    else return device;
}

/**
 * Update firewall information about device
 * @param device given device
 * @param fw true for firewalled device
 * @return true if sucess
 */
global define boolean Change(string device, boolean fw) ``{
    y2debug("Change(%1,%2)", device, fw);
    string dev = device2fw(device);
    if(dev == nil || dev == "" || fw == nil) return false;

    if(fw) {
	if(contains(devices, dev)) {
	    y2error("device already present: %1", device);
	    return false;
	}
	devices = add(devices, dev);
    }
    else {
	if(!contains(devices, dev)) {
	    y2error("device not found: %1", device);
	    return false;
	}
	devices = filter(`l, devices, ``(l!=dev));
    }
    y2debug("devices=%1", devices);
    return true;
}

/**
 * Check if the device is firewalled
 * @param device given device
 * @return true if firewalled
 */
global define boolean Query(string device) ``{
    string dev = device2fw(device);
    if(dev == nil || dev == "") return false;
    return contains(devices, dev);
}

/* EOF */
}
