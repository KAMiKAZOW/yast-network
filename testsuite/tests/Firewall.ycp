{

include "testsuite.ycp";

map R = $[
    "init" : $[
	"scripts" : $[
	    "exists" : true,
	]
    ],
    "probe" : $[
	"system" : [],
    ],
    "target" : $[
	"size" : 1,
	"string" : "Blah\n",
	"tmpdir" : "/tmp",
    ],
];
map E = $[
    "target" : $[
	"bash_output" : $["exit":0],
    ],
];
TESTSUITE_INIT([R],nil);

import "Firewall";

define list FWTEST(list<string> devs, any op) ``{
    Firewall::devices = devs;
    eval(op);
    return Firewall::devices;
}

DUMP("Change(true)");
TEST(``(FWTEST([],``(Firewall::Change("eth0",true)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth1"],``(Firewall::Change("eth0",true)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth0"],``(Firewall::Change("eth0",true)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth0"],``(Firewall::Change("ppp1",true)))), [R,$[],E], nil);
TEST(``(FWTEST(["modem"],``(Firewall::Change("ppp1",true)))), [R,$[],E], nil);

DUMP("Change(false)");
TEST(``(FWTEST([],``(Firewall::Change("eth0",false)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth0"],``(Firewall::Change("eth0",false)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth0", "eth1"],``(Firewall::Change("eth0",false)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth1"],``(Firewall::Change("eth0",false)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth1"],``(Firewall::Change("ppp1",false)))), [R,$[],E], nil);
TEST(``(FWTEST(["eth1", "modem"],``(Firewall::Change("ppp1",false)))), [R,$[],E], nil);
TEST(``(FWTEST(["modem"],``(Firewall::Change("ppp1",false)))), [R,$[],E], nil);

DUMP("Query");
TEST(``(FWTEST([],``(Firewall::Query("eth0")))), [R,$[],E], nil);
TEST(``(FWTEST(["eth0"],``(Firewall::Query("eth0")))), [R,$[],E], nil);
TEST(``(FWTEST(["eth1"],``(Firewall::Query("eth0")))), [R,$[],E], nil);
TEST(``(FWTEST(["modem"],``(Firewall::Query("eth0")))), [R,$[],E], nil);
TEST(``(FWTEST(["modem"],``(Firewall::Query("ppp1")))), [R,$[],E], nil);
TEST(``(FWTEST(["eth0", "modem"],``(Firewall::Query("ppp1")))), [R,$[],E], nil);
TEST(``(FWTEST(["eth0"],``(Firewall::Query("ppp1")))), [R,$[],E], nil);

DUMP("Write");
Firewall::devices = [ "eth0", "ippp1" ];
TEST(``(Firewall::Write()), [R,$[],E], nil);
Firewall::devices = [ "no", "ippp1" ];
TEST(``(Firewall::Write()), [R,$[],E], nil);

map READ = $[
    "init" : $[
	"scripts" : $[
	    "exists" : true,
	    "runlevel" : $[
		"SuSEfirewall2_init" : $["start":["5"],"stop":["5"]],
		"SuSEfirewall2_setup" : $["start":["5"],"stop":["5"]],
		"SuSEfirewall2_final" : $["start":["5"],"stop":["5"]],
	    ]
	]
    ],
    "sysconfig" : $[
	"SuSEfirewall2" : $[
	    "FW_DEV_EXT" : "eth1 ippp0",
	]
    ],
    "probe" : $[
	"system" : [],
    ],
    "target" : $[
	"size" : 1,
	"string" : "Blah\n",
    ],
];

map EXECUTE = $[
    "target" : $[
	"bash_output" : $["exit":0],
    ],
];

DUMP("Read");
TEST(``(Firewall::Read()), [READ, $[], EXECUTE], nil);
DUMP(sformat("Firewall::devices %1",Firewall::devices));

}
