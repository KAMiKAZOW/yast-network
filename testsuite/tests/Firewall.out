Read	.target.tmpdir "/tmp"
Read	.probe.architecture nil
Read	.probe.has_apm nil
Read	.probe.has_pcmcia nil
Read	.probe.has_smp nil
Read	.probe.system []
Dump	Change(true)
Return	["eth0"]
Return	["eth1", "eth0"]
Log	device already present: eth0
Return	["eth0"]
Return	["eth0", "modem"]
Log	device already present: ppp1
Return	["modem"]
Dump	Change(false)
Log	device not found: eth0
Return	[]
Return	[]
Return	["eth1"]
Log	device not found: eth0
Return	["eth1"]
Log	device not found: ppp1
Return	["eth1"]
Return	["eth1"]
Return	[]
Dump	Query
Return	[]
Return	["eth0"]
Return	["eth1"]
Return	["modem"]
Return	["modem"]
Return	["eth0", "modem"]
Return	["eth0"]
Dump	Write
Write	.sysconfig.personal-firewall.REJECT_ALL_INCOMING_CONNECTIONS "eth0 ippp1" true
Write	.sysconfig.personal-firewall nil true
Read	.init.scripts.exists "network" true
Read	.init.scripts.exists "network" true
Read	.init.scripts.runlevel "network" nil
Read	.init.scripts.comment "network" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/network" $["exit":0]
Read	.init.scripts.exists "SuSEfirewall2_init" true
Read	.init.scripts.exists "SuSEfirewall2_init" true
Read	.init.scripts.runlevel "SuSEfirewall2_init" nil
Read	.init.scripts.comment "SuSEfirewall2_init" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/SuSEfirewall2_init" $["exit":0]
Read	.init.scripts.exists "SuSEfirewall2_setup" true
Read	.init.scripts.exists "SuSEfirewall2_setup" true
Read	.init.scripts.runlevel "SuSEfirewall2_setup" nil
Read	.init.scripts.comment "SuSEfirewall2_setup" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/SuSEfirewall2_setup" $["exit":0]
Read	.init.scripts.exists "SuSEfirewall2_final" true
Read	.init.scripts.exists "SuSEfirewall2_final" true
Read	.init.scripts.runlevel "SuSEfirewall2_final" nil
Read	.init.scripts.comment "SuSEfirewall2_final" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/SuSEfirewall2_final" $["exit":0]
Execute	.target.bash "/sbin/SuSEfirewall2 start" 0
Return	true
Write	.sysconfig.personal-firewall.REJECT_ALL_INCOMING_CONNECTIONS "ippp1" true
Write	.sysconfig.personal-firewall nil true
Read	.init.scripts.exists "network" true
Read	.init.scripts.exists "network" true
Read	.init.scripts.runlevel "network" nil
Read	.init.scripts.comment "network" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/network" $["exit":0]
Read	.init.scripts.exists "SuSEfirewall2_init" true
Read	.init.scripts.exists "SuSEfirewall2_init" true
Read	.init.scripts.runlevel "SuSEfirewall2_init" nil
Read	.init.scripts.comment "SuSEfirewall2_init" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/SuSEfirewall2_init" $["exit":0]
Read	.init.scripts.exists "SuSEfirewall2_setup" true
Read	.init.scripts.exists "SuSEfirewall2_setup" true
Read	.init.scripts.runlevel "SuSEfirewall2_setup" nil
Read	.init.scripts.comment "SuSEfirewall2_setup" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/SuSEfirewall2_setup" $["exit":0]
Read	.init.scripts.exists "SuSEfirewall2_final" true
Read	.init.scripts.exists "SuSEfirewall2_final" true
Read	.init.scripts.runlevel "SuSEfirewall2_final" nil
Read	.init.scripts.comment "SuSEfirewall2_final" nil
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/SuSEfirewall2_final" $["exit":0]
Execute	.target.bash "/sbin/SuSEfirewall2 start" 0
Return	true
Dump	Read
Read	.target.size "/etc/sysconfig/personal-firewall" 1
Read	.sysconfig.personal-firewall.REJECT_ALL_INCOMING_CONNECTIONS "eth1 ippp0"
Return	true
Dump	Firewall::devices ["eth1", "ippp0"]
